### ISSUE TITLE
Performance degradation for large results

### ISSUE THREAD

**Issue by coells** [joblib/joblib#1309]
Hello,

I have enountered very bad performance when using `Parallel` to transfer large data from child processes.

```
import numpy as np
import pickle
from time import perf_counter
from joblib import Parallel, delayed


def job1():
    return np.empty(10 ** 9, dtype=np.int8)

def job2():
    x = np.empty(10 ** 9, dtype=np.int8)
    x = pickle.dumps(x)
    x = pickle.loads(x)

if __name__ == '__main__':

    n = 10

    t = perf_counter()
    res = Parallel(n)([delayed(job1)() for _ in range(n)])
    print(perf_counter() - t)

    t = perf_counter()
    for _ in range(n):
        job2()
    print(perf_counter() - t)
```

This example reports `30s` for the first (parallel) loop and `20s` for the second (serial) loop that simulates pickling between processes.

When I try to send the same data between children processes, the performance is as expected, hence this seems to be a problem of Executor when receiving data.

Our real use-case exhibits even worse scenarios, like runtime of 60 minutes instead of 10 minutes as expected.

`Python 3.8.10`
`joblib                 1.1.0`

Thanks,
Tomas

**Comment by ogrisel** [comment]
How many physical cores do you have on this machine? Which OS is this?

I wouldn't expect any speed-up from using parallelism in this case, maybe the performance degradation is expected because the OS has to start Python processes / interpreters concurrently to streaming large chunks of data between processes.

To get a clearer picture of whether there is specific degradation linked to joblib or loky, could you please extend your benchmark to compare against:

- `concurrent.futures.ProcessPoolExecutor(max_workers=n)`;
- `loky.get_reusable_executor(max_workers=n)`.

Could you also try to compare the speed of `pickle.dumps/loads` with cloudpickle.dumps/loads? Maybe you can force the use of `protocol=5` in the calls to dumps.

**Comment by coells** [comment]
_How many physical cores do you have on this machine? Which OS is this?_

36 physical (up to 56 physical), 72 virtual; Centos7

_I wouldn't expect any speed-up from using parallelism in this case_

Do not confuse my example with the actual use-case. We're processing tens to hundreds GB of data in tens of concurrent processes. While it is not usual, it may happen that child sends back to parent up to GB of data, which is were I found perf. issue.

_To get a clearer picture of whether there is specific degradation linked to joblib or loky, could you please extend your benchmark to compare against:_

I can see 10% speed-up for ProcessPoolExecutor and 20% speed-up for loky executor. Though I do not believe this is truly representative as the example above is just a simple show case proving degradation under (almost) no-op circumstance.

Is there a specific goal you are looking for?

Personally, I would believe there's a problem in Parallel class when prioritizing serialization of data from child to parent since running joblib on bigdata is not a usual case? ;) Well, works well until you try to send large data. I may try to switch to loky pool directly to see the real-case difference.

**Comment by coells** [comment]
Hello,

I did more investigation and the problem seems to be due to heavy memory allocation and likely process scheduling order. Parallel/loky only seem to be slow becase execution returns list of results, not an iterator. 

If multiprocessing.Pool is used with effective waiting for results based on callbacks, it is up to 40% faster (in a model case, on large data I saw 60% speedup and more). On the other hand, multiprocessing.Pool with waiting for full completion is only about 20% faster, which might be given just by a chance.

In advance, loky Executor shows the same speeds as Parallel and is even faster than ProcessPoolExecutor by 10%

Anyways, I think I can close this issue.

Thanks,
Tomas
